### PR TITLE
Automated Releases

### PR DESCRIPTION
**Pull request by miraclx**
This introduces a CI workflow that automates GitHub releases as well as publishing to crates.io.

### How to update versions

- Submit a PR that bumps the version of the `Cargo.toml` file.
- Ensure the `Cargo.lock` file is also updated.
- Move all entries under the `Unreleased` section of the `CHANGELOG.md` file to a versioned section.
- Once the PR is merged, the release happens!

### How it works

Unlike most workflows, this doesn't trigger on a tag push. Instead, the first job in the workflow checks for a version bump to the `Cargo.toml` file when merged to master.

If a version bump was detected, we check if that version has already been published to crates.io. If it hasn't, we publish the crate. The second part of this, is checking if the tag already exists on GitHub. If it doesn't, we proceed with the release process.

The release process builds binaries for multiple platforms, and makes them available for download from the final release page as archives, including hashes of each archive in the release note, in case users need to verify them. This also tags the contributors that made the release happen, links to the version of the crate on crates.io, and appropriately tags pre-release versions on GitHub.

### Now for some cases

- If the version has already been published to crates.io and released to GitHub
   CleanShot 2022-11-09 at 13 38 15@2x
- Final release page
   CleanShot 2022-11-09 at 11 59 37@2x
- File Hashes
   CleanShot 2022-11-09 at 11 59 59@2x
- Pre-release tag
   CleanShot 2022-11-09 at 12 28 55@2x

### CODE REVIEW

File: .github/workflows/release.yml [hunk 1]
```diff
@@ -0,0 +1,322 @@
...
+    # ? - { name: 'linux-arm-gnu'     , target: arm-unknown-linux-gnueabihf , os: ubuntu-latest , use-cross: true }
+    # ? - { name: 'linux-arm-musl'    , target: arm-unknown-linux-musleabihf, os: ubuntu-latest , use-cross: true }
+    # ? - { name: 'linux-i686-gnu'    , target: i686-unknown-linux-gnu      , os: ubuntu-latest , use-cross: true }
+    # ? - { name: 'linux-i686-musl'   , target: i686-unknown-linux-musl     , os: ubuntu-latest , use-cross: true }
+    # ? - { name: 'win32-msvc'        , target: i686-pc-windows-msvc        , os: windows-latest,                 }
+        - { name: 'linux-x86_64-gnu'  , target: x86_64-unknown-linux-gnu    , os: ubuntu-latest ,                 }
+        - { name: 'linux-x86_64-musl' , target: x86_64-unknown-linux-musl   , os: ubuntu-latest , use-cross: true }
+        - { name: 'macos-x86_64'      , target: x86_64-apple-darwin         , os: macos-latest  ,                 }
+        - { name: 'win64-x86_64-gnu'  , target: x86_64-pc-windows-gnu       , os: windows-latest,                 }
+        - { name: 'win64-x86_64-msvc' , target: x86_64-pc-windows-msvc      , os: windows-latest,                 }
```
Comment (miraclx) [inline .github/workflows/release.yml:163]: This leads to https://github.com/near/cargo-near/issues/4, I haven't tested `cargo-near` builds on Windows lately.

File: .github/workflows/release.yml [hunk 2]
```diff
@@ -0,0 +1,315 @@
...
+      PROJECT_NAME: ${{ needs.check.outputs.project_name }}
+
+    steps:
+    - name: Checkout repository
+      uses: actions/checkout@v3
+
+    - name: Install rust
+      uses: dtolnay/rust-toolchain@stable
+
+    - uses: Swatinem/rust-cache@v2
```
Comment (austinabell) [inline .github/workflows/release.yml:122]: cache seems unnecessary for the publish step

File: .github/workflows/release.yml [hunk 3]
```diff
@@ -0,0 +1,315 @@
...
+
+        if echo "${RELEASED_VERSIONS}" | grep -q "^v${PROJECT_VERSION}$"; then
+          echo "Project version [${PROJECT_VERSION}] has already been released."
+          echo "SHOULD_RELEASE_BINARY=false" >> $GITHUB_ENV
+        else
+          echo "Project version [${PROJECT_VERSION}] has not been released."
+          echo "SHOULD_RELEASE_BINARY=true" >> $GITHUB_ENV
+        fi
+
+  crate:
```
Comment (austinabell) [inline .github/workflows/release.yml:108]: seems like this should be named `publish`?

**Review by austinabell — APPROVED** [review]

File: .github/workflows/release.yml [hunk 4]
```diff
@@ -0,0 +1,315 @@
...
+
+    - uses: actions/upload-artifact@v3
+      with:
+        name: ${{ env.PKG_PATH }}
+        path: ${{ env.PKG_PATH }}
+        retention-days: 1
+
+  github:
+    runs-on: ubuntu-latest
+    needs: [check, binary]
```
Comment (miraclx) [inline .github/workflows/release.yml:217]: Worth highlighting that the release on GitHub doesn't depend on the crate release.

Yet it includes the crate link in the release notes.

What do you think? Should we wait for the crate to be published, or not?

File: .github/workflows/release.yml (hunk 2 above)
Comment (miraclx) [reply .github/workflows/release.yml]: Indeed, it was helpful for testing, but unnecessary here!

**Comment by itegulov** [comment]
This is very nice. Feels like it deserves to be its own independent Github Action. @miraclx have you considered this?

Also, the release process is valuable knowledge, so lets put it in some visible place like `RELEASE.md`.

**Pull request merged**